src/plugins/QuarterMaster.cpp

```diff
@@ -53,87 +53,98 @@ float CalcSupplies::operator()(float sum, const sc2::Unit* unit_) const {
+
+            case sc2::UNIT_TYPEID::TERRAN_FACTORY:
+            case sc2::UNIT_TYPEID::TERRAN_FACTORYTECHLAB:
+            case sc2::UNIT_TYPEID::TERRAN_FACTORYREACTOR:
+            case sc2::UNIT_TYPEID::TERRAN_STARPORT:
+            case sc2::UNIT_TYPEID::TERRAN_STARPORTTECHLAB:
+            case sc2::UNIT_TYPEID::TERRAN_STARPORTREACTOR:
+            case sc2::UNIT_TYPEID::ZERG_LAIR:
+            case sc2::UNIT_TYPEID::ZERG_HIVE:
+            case sc2::UNIT_TYPEID::PROTOSS_GATEWAY:
+            case sc2::UNIT_TYPEID::PROTOSS_ROBOTICSFACILITY:
+            case sc2::UNIT_TYPEID::PROTOSS_STARGATE:
+                return sum + 2.0f;  // just add 2.0 for all others for now
 
         default:
             return sum;
     }
 }
 
-struct CalcConsumption {
-    float operator()(float sum, const Order& order_) const;
-};
-
-float CalcConsumption::operator()(float sum, const Order& order_) const {
-    return sum + order_.food_required;
-}
 
 }  // namespace
 
 QuarterMaster::QuarterMaster():
-    Plugin(), m_skip_turn(false) {
+    Plugin(), m_skip_until_turn(0) {
 }
 
 void QuarterMaster::OnStep(Builder* builder_) {
-    if (m_skip_turn)
+    if (m_skip_until_turn  > gAPI->observer().GetGameLoop())
         return;
 
     auto units = gAPI->observer().GetUnits();
     std::list<Order> orders = builder_->GetOrders();
 
-    float expected_consumption =
-        gAPI->observer().GetFoodUsed()
-        + 8.0f  // NOTE (alkurbatov): Plan ahead.
-        + std::accumulate(
-            orders.begin(),
-            orders.end(),
-            0.0f,
-            CalcConsumption());
+    if (!orders.empty()
+        && orders.begin()->unit_type_id == gHub->GetCurrentSupplyType()) {
+        m_skip_until_turn = gAPI->observer().GetGameLoop() + 10*22;
+        return;  // wait 10 seconds if already scheduled Supply
+    }
+
+    float expected_consumption = gAPI->observer().GetFoodUsed()
+        + std::accumulate(units().begin(), units().end(), 0.0f, CalcConsumptionRate());
 
     float expected_supply =
-        std::accumulate(units().begin(), units().end(), 0.0f, CalcSupplies())
-        + std::accumulate(
-            orders.begin(),
-            orders.end(),
-            0.0f,
-            CalcSupplies());
+        std::accumulate(units().begin(), units().end(), 0.0f, CalcSupplies());
 
     if (expected_supply > expected_consumption || expected_supply >= 200.0f)
         return;
 
     gHistory.info() << "Request additional supplies: " <<
         expected_consumption << " >= " << expected_supply << std::endl;
 
-    m_skip_turn = true;
-
-    switch (gHub->GetCurrentRace()) {
-        case sc2::Race::Terran:
-            builder_->ScheduleObligatoryOrder(sc2::UNIT_TYPEID::TERRAN_SUPPLYDEPOT, true);
-            return;
-
-        case sc2::Race::Zerg:
-            builder_->ScheduleObligatoryOrder(sc2::UNIT_TYPEID::ZERG_OVERLORD, true);
-            return;
+    m_skip_until_turn = gAPI->observer().GetGameLoop() + 10*22;
```

Comment:
What is the reason to use exactly this magic number? Perhaps it should set according to count of frames skipped between steps?

Reply:
Ok, created constant m_frames_to_skip.